Add clicks

Claudenw opened the pull request:
This pull request adds the ability to explicitly define more CLICK_n and HOLD_nS entries, without resorting to a single script that parses the click or hold times.

# CLICK Changes:

- Clicks CLICK_1 to CLICK_99 may be defined
- Undefined CLICK_n entries are routed to CLICK_OTHER as before
- CLICK_n  config file entries without a script definition:
  - Do not generate an error.
  - Does print an entry indicating the CLICK_n that does not have the script.
  - Does not may any system call to run a script.
- Legacy default CLICK_1, CLICK_2 CLICK_3 entries are preserved.
- If CLICK_OTHER is not defined and CLICK_n is not defined then the result is the same as CLICK_n being defined without a script entry in the config file.

# HOLD changes

- Holds HOLD_1S to HOLD_99S may be defined in odd increments (eg. HOLD_1S and HOLD_3S but not HOLD_2S)
- Undefined HOLD_nS entries are routed to HOLD_OTHER as before
- HOLD_nS config file entries without a script definition:
  - Do not generate an error.
  - Does print an entry indicating the HOLD_n that does not have the script.
  - Does not may any system call to run a script.
- Legacy default HOLD_3S and HOLD_5S entries are preserved,
- if HOLD_OTHER is not defined and HOLD_nS is not defined then the result is the same as HOLD_nS being defined without a script entry in the config file.

# UP/DOWN changes

- UP and DOWN events no longer have a default script.  They must be specified in the configuration file.

# Explicit limits

 - Explicit CLICK_n and HOLD_nS limits are defined at the top of the file.

gtrainavicius commented on pisound-btn/pisound-btn.c:
```diff
@@ -61,14 +61,10 @@ enum action_e
-static const char *const CLICK_1_VALUE_NAME        = "CLICK_1";
-static const char *const CLICK_2_VALUE_NAME        = "CLICK_2";
-static const char *const CLICK_3_VALUE_NAME        = "CLICK_3";
+static const char *CLICK_NAME="CLICK_%u";
```
Please prefer using the same type for global constants as in the other lines - the additional const makes this pointer non-assignable.

Claudenw replied:
changed to
`static const char *const CLICK_NAME="CLICK_%u";`

gtrainavicius commented on pisound-btn/pisound-btn.c:
```diff
@@ -79,27 +75,29 @@ static const char *const PISOUND_VERSION_FILE      = "/sys/kernel/pisound/versio
 
 static const char *const UPDATE_CHECK_DISABLE_FILE = BASE_PISOUND_DIR "/disable_update_check"; // If the file exists, the update check will be disabled.
 
-static const char *const DEFAULT_DOWN              = BASE_SCRIPTS_DIR "/system/down.sh";
```
The default scripts should not be removed, especially for down and up, as it will break LED blinking when the button is down.

Claudenw replied:
Restored defaults

gtrainavicius commented on pisound-btn/pisound-btn.c:
```diff
@@ -79,27 +75,29 @@ static const char *const PISOUND_VERSION_FILE      = "/sys/kernel/pisound/versio
+// must hold the CLICK_OTHER_VALUE_NAME or HOLD_OTHER_VALUE_NAME values
+#define ACTION_NAME_SIZE 11
+// printf( CLICK_%u", ABSOLUTE_MAX_CLICK) must fit in ACTION_NAME_SIZE
+#define ABSOLUTE_MAX_CLICK 99
```
I think it'd be great to simply try reading concrete value like "HOLD_7S", then if such key in the config was not found, then fallback to HOLD_OTHER. This way setting the maximum won't be necessary.

Claudenw replied:
What I wanted was a way to specify an arbitrary HOLD_xS or CLICK_x in the button config without having to go back into the code and put in lots of case or if statements.  I can imagine anyone wanting more than 99 clicks but if so then changing the ABSOLUTE_MAX_CLICK provides a single place to do that.  The actual upper limit is 99,999.  Beyond that the ACTION_NAME_SIZE must be extended.

gtrainavicius commented on pisound-btn/pisound-btn.c:
```diff
@@ -79,27 +75,29 @@ static const char *const PISOUND_VERSION_FILE      = "/sys/kernel/pisound/versio
+#define ABSOLUTE_MAX_HOLD 99
+// convert ticks to seconds so the [0,2) == 1, [2,4) == 3, [4,6) == 5
+#define TICK_2_SECONDS(x) (x/1000 + (x%2))
+static char ACTION_NAME[ACTION_NAME_SIZE+1];
```
Please don't use a global variable here for this, it limits the flexibility of the code - prefer copying into stack variable like get_action_script does.

gtrainavicius commented on pisound-btn/pisound-btn.c:
```diff
@@ -126,7 +124,7 @@ static bool read_line(FILE *f, char *buffer, size_t n)
 }
 
-static void read_config_value(const char *conf, const char *value_name, char *dst, char *args, size_t n, const char *default_value)
+static bool read_config_value(const char *conf, const char *value_name, char *dst, char *args, size_t n, const char *default_value)
```
Please use an int "errno" value instead of a bool, it gives possibility to provide more insight into what went wrong, like there was an invalid argument, or there was some I/O error. Same for get_action_name and get_action_script.

Such errno returns are standard in unix, the 0 always means success, negative values indicate errors, see errno.h for their codes and meanings. They are declared as positive numbers, so you must remember to negate them before returning. Positive values can be used to indicate some success result value, like number of bytes available to read, etc...

Claudenw replied:
converted `read_config_value` back to a `void` type.

gtrainavicius commented on pisound-btn/pisound-btn.c:
```diff
@@ -231,96 +229,141 @@ static void read_config_value(const char *conf, const char *value_name, char *ds
-	read_config_value(g_config_path, action_name, script, arguments, MAX_PATH_LENGTH, default_script);
+	if (!get_action_script(action, ACTION_NAME, action_script, arguments, MAX_PATH_LENGTH ))
+	{
+		return 0;
```
0 should mean success, if get_action_script was returning an errno int itself, you should just return it here too. Otherwise, some relevant negative number for the actual fault from errno.h could be used.

Claudenw replied:
converted `get_action_script` to a `void` type.

gtrainavicius commented on pisound-btn/pisound-btn.c:
```diff
@@ -330,42 +373,47 @@ static int get_action_script_path(enum action_e action, unsigned arg0, unsigned
 
+	if (n == 0)
+	{
+		fprintf(stderr, "execute_action: no command for action %s : click count %u hold time %u (%u seconds)\n", ACTION_NAME, click_count, hold_time, TICK_2_SECONDS(hold_time));
```
This block won't be necessary once return value is fixed. The above line could be expanded to include more information like click & hold time.

Claudenw replied:
It will still be necessary, except the test will be a little different.  This is because the config now allows you to specify an action without a command.

So if I specify
```
ON_UP
ON_DOWN  /some/script/file
```

The ON_UP will no longer trigger the default.  I provides a way to cancel the command without overriding it.

Claudenw replied:
changed `(n ==0)` to `( cmd[0]='\0')` as this indicates that there is no script to execute.
Added more detail to the original message.

gtrainavicius commented:
Hey, thank you for sending this pull request. I've quickly went through the code and identified a couple of things to fix.

Also, there's a lot of commits in the pull request that should simply just be 'squashed' (merged) into 1 or a couple and the commit message should provide all the necessary information to know what the changes are about. :)

Basically this keeps only the relevant information for the history of the code, so it's easier for everyone to see what's going on.

Claudenw commented:
I think I managed to remove the excess comments.  I probably ought to edit the commit messages now to declare what this change does to the overall operation, not be specific to the internals of this change.

gtrainavicius commented:
Ok, I've looked through the latest code, looks like all the things are solved. I'll test it out some more and merge once the final commit(s) & their messages are fixed up. :) I'd like the commit messages to state what behavior changed.

Claudenw commented:
I think the message change is correct now.  Does it meet with your approval?

Claude

On Mon, Feb 15, 2021 at 1:37 PM gtrainavicius <notifications@github.com>
wrote:

> Ok, I've looked through the latest code, looks like all the things are
> solved. I'll test it out some more and merge once the final commit(s) &
> their messages are fixed up. :) I'd like the commit messages to state what
> behavior changed.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/BlokasLabs/pisound/pull/12#issuecomment-779229471>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AASTVHQM2LZQA2VD2C6GKG3S7EPRRANCNFSM4W3HJWLQ>
> .
>


-- 
I like: Like Like - The likeliest place on the web
<http://like-like.xenei.com>
LinkedIn: http://www.linkedin.com/in/claudewarren

gtrainavicius commented on pisound-btn/pisound-btn.c:
```diff
@@ -330,42 +380,48 @@ static int get_action_script_path(enum action_e action, unsigned arg0, unsigned
 	args[n-1] = '\0';
 
-	return strlen(dst);
+	return 0;
```
It should return the length of the string as it did before, as it's used by the calling function, to know where to append the click / hold arguments. Otherwise, it tries to invoke " 1" (just the arguments) for example for single click.

gtrainavicius commented on pisound-btn/pisound-btn.c:
```diff
@@ -330,42 +380,48 @@ static int get_action_script_path(enum action_e action, unsigned arg0, unsigned
 		return;
 	}
 
+	if (cmd[0] == '\0')
```
It'd be better to just check for n == 0 to know there's no action.

gtrainavicius commented on pisound-btn/pisound-btn.c:
```diff
@@ -233,94 +232,145 @@ static void read_config_value(const char *conf, const char *value_name, char *ds
 
+	if (action_script[0] == '\0')
+	{
+		return 0;
```
It should set at least script[0] to '\0', to avoid having uninitialized output variable

gtrainavicius reviewed (changes requested):
Testing this, looks clicks don't actually call the script, most likely reason identified in comments on code.

Claudenw commented:
I have updated the pull request.

I added a debug() method that replaces sprintf( stderr, .... ) for some
statements.  And added more logging as well as --debug and -q options.  At
default setting the output from the earlier code is unchanged.  "--debug 0"
or "-q" removes all but true error messages.  "--debug 2" or higher adds
more debugging. (it gets noisy very quickly)

I am attaching a tgz with some simple test scripts (they just print the
name and arguments) as well as a pisound.conf to show all the options in
the new code.



On Thu, Feb 18, 2021 at 10:41 AM gtrainavicius <notifications@github.com>
wrote:

> *@gtrainavicius* requested changes on this pull request.
>
> Testing this, looks clicks don't actually call the script, most likely
> reason identified in comments on code.
> ------------------------------
>
> In pisound-btn/pisound-btn.c
> <https://github.com/BlokasLabs/pisound/pull/12#discussion_r578293013>:
>
> >  	}
>
>  	strncpy(args, arguments, n-1);
>  	args[n-1] = '\0';
>
> -	return strlen(dst);
> +	return 0;
>
> It should return the length of the string as it did before, as it's used
> by the calling function, to know where to append the click / hold
> arguments. Otherwise, it tries to invoke " 1" (just the arguments) for
> example for single click.
> ------------------------------
>
> In pisound-btn/pisound-btn.c
> <https://github.com/BlokasLabs/pisound/pull/12#discussion_r578293500>:
>
> >  		return;
>  	}
>
> +	if (cmd[0] == '\0')
>
> It'd be better to just check for n == 0 to know there's no action.
> ------------------------------
>
> In pisound-btn/pisound-btn.c
> <https://github.com/BlokasLabs/pisound/pull/12#discussion_r578293950>:
>
> >
> +	if (action_script[0] == '\0')
> +	{
> +		return 0;
>
> It should set at least script[0] to '\0', to avoid having uninitialized
> output variable
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/BlokasLabs/pisound/pull/12#pullrequestreview-593051525>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AASTVHS6S7YZ3U6YJPTFNXLS7TVGJANCNFSM4W3HJWLQ>
> .
>


-- 
I like: Like Like - The likeliest place on the web
<http://like-like.xenei.com>
LinkedIn: http://www.linkedin.com/in/claudewarren

Claudenw commented:
Are there any remaining issues with this pull request?

gtrainavicius commented:
Ok, we've released version 1.12 on our apt server.
